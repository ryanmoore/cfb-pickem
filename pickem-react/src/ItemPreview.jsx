
import React, { PropTypes, Component }  from 'react';
import DragLayer from 'react-dnd/lib/DragLayer';

const layerStyles = {
    position: 'fixed',
    pointerEvents: 'none',
    zIndex: 100,
    left: 0,
    top: 0,
    width: '100%',
    height: '100%'
};

function collect(monitor) {
    return {
        currentOffset: monitor.getSourceClientOffset(),
        isDragging: monitor.isDragging()
    };
}

function stylePreview(offset) {
    // Hide if not dragging
    if(!offset) {
        return { display: 'none' };
    }

    const transform = `translate(${offset.x}px, ${offset.y}px)`;
    return { pointerEvents: 'none',
        transform: transform,
        WebkitTransform: transform
    };
}

class ItemPreview extends Component {
    static propTypes = {
        currentOffset: PropTypes.shape({
            x: PropTypes.number.isRequired,
            y: PropTypes.number.isRequired,
        }),
        isDragging: PropTypes.bool.isRequired
    };

    render() {
        const { currentOffset, isDragging, children } = this.props;
        if(!isDragging) {
            return null;
        }
        return (
            <div style={ layerStyles }>
                <div style={stylePreview(currentOffset)}>
                    { children }
                </div>
            </div>
        );
    }
}

export default DragLayer(collect)(ItemPreview);
