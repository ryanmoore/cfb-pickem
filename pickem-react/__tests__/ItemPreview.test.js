
import React from 'react';
import ItemPreview from '../src/ItemPreview';
import renderer from 'react-test-renderer';

describe('ItemPreview', () => {
    // Pull the item we wrote out from the Drag-and-Drop wrapper
    const RawItemPreview = ItemPreview.DecoratedComponent;

    it('should be hidden/null when not dragging', () => {
        const component = renderer.create(
            <RawItemPreview isDragging={false}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be hidden/null when not dragging even with known offset', () => {
        const component = renderer.create(
            <RawItemPreview isDragging={false} currentOffset={{x:0, y:0}}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be hidden/null when offset is unknown', () => {
        const component = renderer.create(
            <RawItemPreview isDragging={true} currentOffset={null}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should take the value of the current offset when dragging', () => {
        const component = renderer.create(
            <RawItemPreview isDragging={true} currentOffset={ {x:0, y:0} }/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should transform to match current offset when dragging', () => {
        const component = renderer.create(
            <RawItemPreview isDragging={true} currentOffset={ {x:1, y:2} }/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
