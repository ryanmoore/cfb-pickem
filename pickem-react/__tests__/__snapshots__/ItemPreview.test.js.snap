exports[`ItemPreview should be hidden/null when not dragging 1`] = `null`;

exports[`ItemPreview should be hidden/null when not dragging even with known offset 1`] = `null`;

exports[`ItemPreview should be hidden/null when offset is unknown 1`] = `
<div
  style={
    Object {
      "left": 0,
      "pointerEvents": "none",
      "position": "fixed",
      "top": 0,
      "zIndex": 100,
    }
  }>
  <div
    style={
      Object {
        "display": "none",
      }
    } />
</div>
`;

exports[`ItemPreview should take the value of the current offset when dragging 1`] = `
<div
  style={
    Object {
      "left": 0,
      "pointerEvents": "none",
      "position": "fixed",
      "top": 0,
      "zIndex": 100,
    }
  }>
  <div
    style={
      Object {
        "WebkitTransform": "translate(0px, 0px)",
        "pointerEvents": "none",
        "transform": "translate(0px, 0px)",
      }
    } />
</div>
`;

exports[`ItemPreview should transform to match current offset when dragging 1`] = `
<div
  style={
    Object {
      "left": 0,
      "pointerEvents": "none",
      "position": "fixed",
      "top": 0,
      "zIndex": 100,
    }
  }>
  <div
    style={
      Object {
        "WebkitTransform": "translate(1px, 2px)",
        "pointerEvents": "none",
        "transform": "translate(1px, 2px)",
      }
    } />
</div>
`;
